### ISSUE TITLE
utf-8 character encoding / decoding problem

### ISSUE THREAD

**Issue by GergelyLakatos** [Adldap2/Adldap2-Laravel#524]
- Laravel Version: 5.5
- Adldap2-Laravel Version: 4.0.7
- PHP Version: 7.0.7
- LDAP Type: OpenLDAP

### character encoding / decoding issue.

I have to work with hungarian characters, what is encoded on the LDAP server in UTF-8, but when I get back I got "SzÃ¡sz" (instead of "Szász"). If I use `utf8_decode()` php function I can convert "SzÃ¡sz" to "Szász".

My question, is there any way to set utf-8 caracter coding?

Thanks the responses in advance! :)

**Comment by stevebauman** [comment]
Hi @GergelyLakatos, unfortunately not at the moment, this feature completely slipped my mind.

Will look to implement this via configuration array and have it decode on retrieving attributes from models.

Possible example:

```php
$config = [
    '...',
    'encoding' => 'utf-8',
];

$ad = new Adldap(['default' => $config]);
```

This may be a little tricky to perform though since not all LDAP attributes will be UTF-8 (such as object SID / GUID)...

Thoughts?

**Comment by GergelyLakatos** [comment]
Hello @stevebauman !

First of all, thanks for the quick response. It will be a great feature, can't wait the release!

I did some search about this problem.
The quick solution is setting this: `ldap_set_option($ldap, LDAP_OPT_PROTOCOL_VERSION, 3);`
So it will be a good feature if there would be an array in config, where you can define key value pairs,
e.g.: ["LDAP_OPT_PROTOCOL_VERSION" => 3]. Check [this](https://stackoverflow.com/a/11038533/5349795) answer.

(I check other code of my company, they connect to ldap "native way" (not via lib) and everytime they set this protocol version. )

An another solution that you can define a key value array (besides of encoding), where you can define the encoding of spec attributes. There is a good [answer in stackoverflow](https://stackoverflow.com/a/41060909/5349795) what i suggest you.

**Comment by stevebauman** [comment]
Hi @GergelyLakatos, no problem.

There is actually already a `version` configuration parameter. It just isn't in the default configuration that is published.

All configuration parameters can be set in the `adldap.php` file that is available on the root Adldap2 package:

https://github.com/Adldap2/Adldap2/blob/master/src/Configuration/DomainConfiguration.php#L32

You just need to add the following line to configure this in your `adldap.php` file:

```php
'connection_settings' => [

    //...

    'version' => 3,
```

However the default is supposed to be set to `3`, did you change this in your app somewhere else?

https://github.com/Adldap2/Adldap2/blob/master/src/Connections/Provider.php#L256

https://github.com/Adldap2/Adldap2/blob/master/src/Configuration/DomainConfiguration.php#L31-L32

Also, for the StackOverflow post, this is already included when changing passwords:

https://github.com/Adldap2/Adldap2/blob/0a0fda482f956bd28329580f2907b2b533a2548a/src/Models/User.php#L1076-L1096

**Comment by GergelyLakatos** [comment]
Hi @stevebauman !

It's getting weird. I did not change the version. 
I add to config/adldap.php file:
```
[
....
 'connections' => [
        ....
        'default' => [
            ....
            'connection_settings' => [
                ....
                'version' => 3,
]]]]
```
but there is no change. :( 

My code is a simple query (The auth works fine.):

```
    $result = Adldap::search()
      ->select($select)
      ->rawFilter($searchFilter)
      ->in($userbasedn)
      ->sortBy($sortby, 'asc')
      ->get();
```
The ugly solution, is a recursive function, what checks every variable coding, but its kinda hacking..

**Comment by stevebauman** [comment]
Hi @GergelyLakatos, try clearing your configuration cache using `php artisan config:clear`.

Your second example won't work as the version needs to be set before an `ldap_bind` takes place.

**Comment by GergelyLakatos** [comment]
```
Hi @stevebauman !

I have already run the config clear. I gave a chance for this:

    Adldap::getConnection()->setOption(LDAP_OPT_PROTOCOL_VERSION, 3);
    Adldap::getConnection()->bind(env('ADLDAP_ADMIN_USERNAME'), env('ADLDAP_ADMIN_PASSWORD'));

    $result = Adldap::search()
      ->select($select)
      ->rawFilter($searchFilter)
      ->in($userbasedn)
      ->sortBy($sortby, 'asc')
      ->get();
      Log::debug("getUsers result: ", [$result]);
```

but the problem stays.. 
I will made a new provider, it may help. :)

**Comment by GergelyLakatos** [comment]
Hi!

other odd behaviour:

I made a new Connection:

```
class LDAPConnector extends \Adldap\Connections\Ldap
{
    public function connect($hostname = [], $port = '389')
    {
      // $connections = $this->getConnectionString($hosts, $this->getProtocol(), $port);
      $this->connection = ldap_connect(env('ADLDAP_CONTROLLERS'),env('ADLDAP_PORT'));
      ldap_set_option($this->connection, LDAP_OPT_PROTOCOL_VERSION, 3);
      return $this->connection;
    }
}
```
Then I use this;
```

    $connection = new LDAPConnector();
    $provider = new \Adldap\Connections\Provider([], $connection);

    $result = $provider->search()
      ->select($select)
      ->rawFilter($searchFilter)
      ->in($userbasedn)
      ->sortBy($sortby, 'asc')
      ->get();
      Log::debug("getUsers result: ", [$result]);
```
same result..

but if I run this code:

```
      $ldapconn = ldap_connect(env('ADLDAP_CONTROLLERS'),env('ADLDAP_PORT'));
      ldap_set_option($ldapconn, LDAP_OPT_PROTOCOL_VERSION, 3);
      $ldapbind = ldap_bind($ldapconn, env('ADLDAP_ADMIN_USERNAME'), env('ADLDAP_ADMIN_PASSWORD'));
      $res = ldap_search($ldapconn, $userbasedn, $filterString1, $select);
      $entries = ldap_get_entries($ldapconn, $res);
      Log::debug("getUsers entries: ", [$entries]);
```
Its working fine.. Even if I comment out the LDAP_OPT_PROTOCOL_VERSION line..

**Comment by GergelyLakatos** [comment]
Heureka! My colleague found the problem.

It is [here](https://github.com/Adldap2/Adldap2/blob/fd4d9dbfcabe3a3881fc1491715cf455174b3d16/src/Models/Model.php#L215) in the `Adldap\Models\Model` If I comment out the 215, 216, 217 lines, the problem disappear.

Can you put this into a config (so I can avoid this code running with a config set)?

**Comment by stevebauman** [comment]
Awesome thanks @GergelyLakatos! Appreciate you debugging this issue, will come out with a patch shortly.

**Comment by stevebauman** [comment]
Just to confirm, I'm assuming your consuming your results in an API? Otherwise `jsonSerialize()` will never be called on the model instances.

**Comment by stevebauman** [comment]
Hmm, this actually might not be able to be done, as JSON requires UTF-8 and is not compatible with ISO-8859-1. It's doing the right thing here.

Attributes that are retrieved from your LDAP server are still UTF-8 (this is just the encoding used by LDAP servers) and will need to be converted to ISO-8859-1 manually.

I'll see if I can implement something that can be easily configurable for automatically encoding attributes to a different format upon retrieval (`$user->cn`).

Going to change this back to `enhancement`.

**Comment by stevebauman** [comment]
Hi @GergelyLakatos ,thanks for the response.

> The code should check every variable encoding, and if it is not utf-8 you should encode to utf-8

Yes definitely, I'll patch this now.

> There should be a config param, where I can "turn on" this there lines.

I don't think we should really offer this to be configured, as JSON must be UTF-8, otherwise it is invalid JSON.

> If the LDAP_OPT_PROTOCOL_VERSION is 3 (what means UTF-8 is default), the code should skip this 3 line.

I'll patch this in as well thanks! :)

**Comment by stevebauman** [comment]
After a ton of testing, I've determined that this 100% can't be changed or configured.

When LDAP returns results, it can contain `BINARY` strings. These cannot be used in JSON objects.

`json_encode()` fails entirely when you try to encode any binary, so encoding to `UTF-8` is absolutely necessary so we **at least** return an empty string to guarantee the encoding for all attributes on the Adldap model.

Another problem lies with how binary strings are given in LDAP results, as some may be contained in double quotes which causes more problems.

JSON objects **require** the UTF-8 character set, there's no way around this.

You will need to perform encoding yourself and not rely on the built in `jsonSerialize` method that is implemented on models.

No we could avoid all of this encoding mess by simply restricting what attributes can be returned by the `jsonSerialize()` method, but this causes issues with LDAP types, as not all attributes will exist on all different installations (OpenLDAP, ActiveDirectory, FreeIPA, Sun Directory Server etc.).

You may be able to get it to work by simply removing the `utf8_encode()` method, but it will fail on any object thereafter that contains a `BINARY` string, making the `jsonSerialize()` method on models completely useless and unpredictable.

To encode models yourself, simply run an `array_walk_recursive()` on all the models attributes (`$model->getAttributes()`) to encode them to whatever you'd like.

Attributes are not touched at all in Adldap2 besides when running the `jsonSerialize()` method.

If you are able to figure out how to properly encode *all* `BINARY` strings returned from any LDAP result into a JSON string, let me know.

**Comment by GergelyLakatos** [comment]
Hi!

may I suggest you this update, its working fine for me (if is already utf-8 not encode again):
```
        array_walk_recursive($attributes, function(&$val){
          if(!mb_detect_encoding($val, 'UTF-8', true) ) {
            $val = utf8_encode($val);
          }
        });
```
If I dont want to use this "vendor hack" I have to do this util functions:

```
  public static function utf8decodeArray(&$userEntries){
    for ($i=0; $i < count($userEntries); $i++) {
      $model = $userEntries[$i];
      $attributes = $model->getAttributes();
      self::utf8decodeModel($attributes);
    }
  }

  public static function utf8decodeModel(&$attributes) {
    array_walk_recursive($attributes, function(&$val){
        $val = utf8_decode($val);
    });
  }
```
So, I have to reverse every encoding.. 

Is there any chance to use my recommendation?

**Comment by stevebauman** [comment]
Your suggestion works! Thanks @GergelyLakatos!

I previously tried using `mb_detect_encoding`, but didn't know there was a `$strict` parameter that removes the issue of it detecting binary strings as UTF-8.

I've patched this now in `dev-master`, but will also put this in the latest release.

**Comment by stevebauman** [comment]
Released in `v8.1.5`.

**Comment by GergelyLakatos** [comment]
Thanks! You are welcome. :)
